daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala

```diff
@@ -196,94 +199,109 @@ private[lf] final class ValueTranslator(
+                      // Not fully labelled (or reordering disabled), so order dependent
+                      // Additional fields should downgrade, missing fields should upgrade
+                      case None => {
+                        val correctFields = targetFieldsAndTypes.toSeq.zipWithIndex.map {
+                          case ((lbl, ty), i) =>
+                            val (mbLbl, v) =
+                              sourceElements.get(i).getOrElse(addMissingField(lbl, ty))
+                            mbLbl.foreach(lbl_ =>
+                              if (lbl_ != lbl)
                                 typeError(
-                                  s"Mismatching record field label '$sourceField' (expecting '$targetField') for record $tyCon"
+                                  s"Mismatching record field label '$lbl_' (expecting '$lbl') for record $tyCon"
                                 )
-
-                          }
-                          val typ: Type = AstUtil.substitute(targetFieldType, subst)
-                          val sv: SValue = go(typ, v, newNesting)
-                          List(sv)
-                        case None => // DOWNGRADE
-                          // i ranges from 0 to numS-1. So i >= numT implies numS > numT
-                          assert((numS > i) && (i >= numT))
-                          v match {
-                            case ValueOptional(None) =>
-                              List() // ok, drop
-                            case ValueOptional(Some(_)) =>
-                              typeError(
-                                "An optional contract field with a value of Some may not be dropped during downgrading."
-                              )
-                            case _ =>
-                              typeError(
-                                "Unexpected non-optional extra contract field encountered during downgrading."
-                              )
-                          }
-                      }
-                    }.toList
-
-                  val fields: ImmArray[Ref.Name] =
-                    targetFieldsAndTypes.map { case (name, _) =>
-                      name
-                    }
-
-                  val values: util.ArrayList[SValue] = {
-                    if (numT > numS) {
-
-                      def isOptionalType(typ: Type): Boolean = {
-                        typ match {
-                          case TApp(TBuiltin(BTOptional), _) => true
-                          case _ => false
+                            )
+                            (lbl, v, ty)
                         }
+                        val numS = sourceElements.length
+                        val numT = targetFieldsAndTypes.length
+                        // We have extra fields
+                        val extraFields =
+                          if (numS > numT)
+                            sourceElements.strictSlice(numT, numS)
```

Comment:
I do not think it is a problem at all here, but just be aware that `strictSlice` is linear.

Reply:
Is it? `sourceElements` here is an `ImmArray`, and from what I can see in our implementation, it just makes a new ImmArray with different start/length and the same internal array